nujo/autodiff/_functions/_transform.py

```diff
@@ -53,3 +55,134 @@ def backward(self, idx: int, accum_grad: Function.T) -> Function.T:
+
+        self.padding = padding
+
+    def forward(self) -> ndarray:
+        return pad(self.children[0].value, (
+            (0, 0),
+            (0, 0),
+            (self.padding[0], self.padding[0]),
+            (self.padding[1], self.padding[1]),
+        ))
+
+    def backward(self, idx: int, accum_grad: Function.T) -> Function.T:
+        return accum_grad[:, :, self.padding[0]:-self.padding[0],
+                          self.padding[1]:-self.padding[1]]
+
+
+# ====================================================================================================
+
+
+class _Im2col(Function):
+    ''' Image to column shape transformation
+
+    The local regions in the input image are stretched out into columns.
+
+    For example, if the input is [3x227x227] and it is to be convolved
+    with 3x11x11 filters at stride (4, 4), then we would take [3x11x11]
+    blocks of pixels in the input and stretch each block into a column
+    vector of size 3*11*11 = 363. Iterating this process in the input
+    at stride of (4, 4) gives (227-11)/4+1 = 55 locations along both
+    height and width, leading to an output matrix X_col of Im2col of size
+    [363 x 3025], where every column is a stretched out receptive field
+    and there are 55*55 = 3025 of them in total.
+
+    Reference: CS231n Stanford
+    (https://cs231n.github.io/convolutional-networks/)
+
+    Parameters:
+    -----------
+     - input : image shaped array, shape: (batch_size, channels, height, width)
+     - kernel_size : tuple of 2 integers, image filter height and width
+     - stride : tuple of 2 integers
+
+    '''
+    def __init__(self, input: Union[Tensor, ndarray, List[Number], Number],
+                 kernel_size: Tuple[int, int], stride: Tuple[int, int]):
+
+        super(_Im2col, self).__init__(input)
+
+        # Shape of `input` should be: (batch_size, channels, height, width)
+        assert len(self.children[0].shape) == 4
+
+        self.kernel_size = kernel_size
+
+        # Calculate the indices where the dot products are
+        # to be applied between weights and the image
+        self._im2col_indices: Tuple[ndarray, ndarray, ndarray] =\
+            self._get_im2col_indices(self.kernel_size, stride)
+
+    def forward(self) -> ndarray:
+        ''' Method which turns the image shaped input to column shape
+        '''
+
+        images = self.children[0].value
+
+        # Reshape content into column shape
+        n_features = self.kernel_size[0] * self.kernel_size[1] *\
+            images.shape[1]  # number of channels
+
+        k, i, j = self._im2col_indices
+        return images[:, k, i, j].transpose(1, 2, 0).reshape(n_features, -1)
+
+    def backward(self, idx: int, accum_grad: Function.T) -> Function.T:
+        ''' Method which turns the column shaped input to image shape
+        '''
+
+        images = zeros(self.children[0].shape)
+
+        k, i, j = self._im2col_indices
+        add.at(images, (slice(None), k, i, j),
```

Comment:
couldn't you have used images = add(...)

Reply:
No, `add` is different from `add.at`. See [here](https://numpy.org/doc/stable/reference/generated/numpy.ufunc.at.html).